### PR TITLE
fix(sql): fix reading uint16 column as chars from parquet

### PR DESCRIPTION
**Pull request by ideoma**
UInt16 parquet logical type used to be read as a QuestDB CHAR type, but sometimes it can be a genuine integer value and an invalid character.

This PR saves QuestDB column-type integer values in the parquet file using additional key-value pairs as 

```
__qdb_type__<column name>: int32 column type
```

With this change, the parquet column is read as QuestDB CHAR only when the column was written by QuestDB.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/ReadParquetRecordCursor.java|22|22|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/ReadParquetRecordCursor.java|22|22|100.00%|

File: core/rust/qdbr/src/parquet_write/schema.rs [hunk 1]
```diff
@@ -329,15 +331,29 @@ pub struct Partition {
...
-    Ok(SchemaDescriptor::new(
-        partition.table.clone(),
-        parquet_types,
+
+    let additinal_keyvals = partition
+        .columns
+        .iter()
+        .map(|c| {
+            KeyValue::new(
+                QDB_TYPE_META_PREFIX.to_string() + c.name,
```
Comment (eugenels) [inline core/rust/qdbr/src/parquet_write/schema.rs:348]: use format!("{}{}", QDB_TYPE_META_PREFIX, c.name)

File: core/rust/qdbr/src/parquet_write/schema.rs [hunk 2]
```diff
@@ -329,15 +331,29 @@ pub struct Partition {
...
+                c.full_column_type.to_string(),
```
Comment (eugenels) [inline core/rust/qdbr/src/parquet_write/schema.rs:349]: why not update ColumnType to be "full". update try_from(v: i32) and ask ChatGPT to generate to_string/from_string type safe conversion?

File: core/rust/qdbr/src/parquet_read/meta.rs [hunk 3]
```diff
@@ -199,7 +229,7 @@ mod tests {
 
         for (i, col) in meta.columns.iter().enumerate() {
             let (col_type, _, name) = cols[i];
-            assert_eq!(col.typ, to_storage_type(col_type));
+            assert_eq!(col.typ, col_type);
```
Comment (eugenels) [inline core/rust/qdbr/src/parquet_read/meta.rs:232]: do we have something for ColumnType::Char ?

File: core/rust/qdbr/src/parquet_read/meta.rs [hunk 4]
```diff
@@ -1,30 +1,39 @@
...
-
 impl ParquetDecoder {
     pub fn read(mut reader: File) -> anyhow::Result<Self> {
         let metadata = read_metadata(&mut reader)?;
-
         let col_len = metadata.schema_descr.columns().len();
+        let additional_meta: Option<HashMap<String, Option<String>>> =
+            metadata.key_value_metadata.as_ref().map(|vec| {
+                vec.iter()
+                    .map(|kv| (kv.key.to_string(), kv.value.to_owned()))
```
Comment (eugenels) [inline core/rust/qdbr/src/parquet_read/meta.rs:20]: do we allow keys duplicates? if not, it's good to change API from Vec to HashMap

File: core/rust/qdbr/src/parquet_read/meta.rs [hunk 5]
```diff
@@ -51,8 +60,27 @@ impl ParquetDecoder {
...
+        des: &Descriptor,
+        additional_meta: &Option<HashMap<String, Option<String>>>,
+    ) -> Option<(ColumnType, i32)> {
+        let result = additional_meta.as_ref().and_then(|hashmap| {
+            hashmap.get(&(QDB_TYPE_META_PREFIX.to_string() + &des.primitive_type.field_info.name))
+        });
+        if let Some(Some(val)) = result {
+            let col_type = match val.parse::<i32>() {
+                Ok(col_type_int) => match ColumnType::try_from(col_type_int) {
+                    Ok(col_type) => return Some((col_type, col_type_int)),
```
Comment (eugenels) [inline core/rust/qdbr/src/parquet_read/meta.rs:73]: type unsafe

File: core/rust/qdbr/src/parquet_write/schema.rs (hunk 2 above)
Comment (ideoma) [reply core/rust/qdbr/src/parquet_write/schema.rs:349]: Full column type will include some column bit flags, like geohashes do. Maybe we do it later on but we rely on may places on those int values to be stable

File: core/rust/qdbr/src/parquet_read/meta.rs (hunk 3 above)
Comment (ideoma) [reply core/rust/qdbr/src/parquet_read/meta.rs:232]: I don't understand the question, can you rephrase it?

File: core/rust/qdbr/src/parquet_read/meta.rs (hunk 4 above)
Comment (ideoma) [reply core/rust/qdbr/src/parquet_read/meta.rs:20]: Should be no dups. I'd expect we do use HashMap since:
```
dditional_meta: Option<HashMap<String, Option<String>>>
```

File: core/rust/qdbr/src/parquet_read/meta.rs (hunk 5 above)
Comment (ideoma) [reply core/rust/qdbr/src/parquet_read/meta.rs:73]: Not sure I see type unsafeness here, can you explain?

File: core/src/main/java/io/questdb/griffin/engine/functions/table/ReadParquetRecordCursor.java [hunk 6]
```diff
@@ -261,6 +287,30 @@ public short getShort(int col) {
             return Unsafe.getUnsafe().getShort(dataPtr + currentRowInRowGroup * 2L);
         }
 
+        @Override
+        public CharSequence getStrA(int col) {
+            long auxPtr = auxPtrs.get(col);
```
[the remark this answers is not in the document]
Comment (ideoma) [reply core/src/main/java/io/questdb/griffin/engine/functions/table/ReadParquetRecordCursor.java:292]: It doesn't give much, I tried

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 22 / 22 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/ReadParquetRecordCursor.java|22|22|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 17 / 17 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/table/ReadParquetRecordCursor.java|17|17|100.00%|

File: core/rust/qdbr/src/parquet_read/meta.rs (hunk 5 above)
Comment (ideoma) [reply core/rust/qdbr/src/parquet_read/meta.rs]: We talked about it on the phone, I believe we agreed that it's not unsafe but can be potentially restrictive in some future use and we will improve it then.

File: core/rust/qdbr/src/parquet_read/meta.rs (hunk 4 above)
Comment (ideoma) [reply core/rust/qdbr/src/parquet_read/meta.rs:20]: No duplicates if it's us writing the file.

File: core/rust/qdbr/src/parquet_read/meta.rs (hunk 5 above)
Comment (eugenels) [reply core/rust/qdbr/src/parquet_read/meta.rs]: the `val` value can be not a valid column type encoding.
ok to keep it as it is now.
